llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -0,0 +1,214 @@
+                dist.get_global_rank() == 0
+        ]):
+            self.launch_run()
+
+            self.last_launch = state.timestamp.batch
+            self.count += 1
+
+    def _get_current_run(self) -> Run:
+        if os.environ.get(MOSAICML_PLATFORM_ENV_VAR,
+                          'false').lower() == 'false':
+            raise RuntimeError(
+                'AsyncEval callback is only supported when running on the MosaicML platform'
+            )
+
+        run_name = os.environ.get(RUN_NAME_ENV_VAR, None)
+        if not run_name:
+            raise RuntimeError(
+                'RUN_NAME environment variable must be set to use the AsyncEval callback'
+            )
+
+        # Allows the MapiException to be raised if the run doesn't exist
+        return get_run(run_name, include_details=True)
+
+    def get_eval_parameters(
+        self,
+        parameters: Dict[str, Any],
+        run_name: str,
+    ) -> Dict[str, Any]:
+        looking_for = REQUIRED_PARAMS_FOR_EVAL.copy()
+
+        # Go through all parameters and pull out the ones needed for eval
+        subset_keys = {}
+        for key in parameters:
+            if key in OPTIONAL_PARAMS_FOR_EVAL:
+                subset_keys[key] = parameters[key]
+            elif key in REQUIRED_PARAMS_FOR_EVAL:
+                subset_keys[key] = parameters[key]
+                looking_for.remove(key)
+
+        if looking_for:
+            raise Exception(
+                f'Missing the following required parameters for async eval: {looking_for}'
+            )
+
+        # Convert the save_folder to a load_path
+        subset_keys['load_path'] = get_load_path(
+            subset_keys.pop('save_folder'),
+            parameters.get('save_latest_filename', None))
+
+        # Update the loggers to use the training run name
+        for logger, config in subset_keys.get('loggers', {}).items():
+            if logger == 'wandb':
+                config['name'] = config.get('name', run_name)
+            elif logger == 'mlflow':
+                config['run_name'] = config.get('run_name', run_name)
+
+        # Create new eval models list
+        subset_keys['models'] = get_eval_models_dict(
+            subset_keys.pop('model'), subset_keys.pop('tokenizer'))
+
+        subset_keys['run_name'] = get_run_name(run_name, 0)
+        return subset_keys
+
+    def launch_run(self) -> Run:
+        cfg = self.current_run.submitted_config
+        default_compute = {
+            'gpus': 8,
+            'cluster': self.current_run.cluster,
+        }
+        params = self.get_eval_parameters(cfg.parameters or {},
+                                          self.current_run.name)
+
+        # TODO: This just runs an eval run, but we also want to attach the
+        # deployment, which would require a hf conversion and parametrizing the
+        # dependent_deployment in the run config
+        command = 'cd llm-foundry/scripts \n composer eval/eval.py $PARAMETERS'
```

Comment:
I suppose this doesn't work if I don't use the parameters section. It also doesn't work if I command line override any of the parameters. Should we be using the config that gets passed to `train.py` instead?

Reply:
Updated in b58ccf9c33d6d1b3227762fb7a8c1552a350a56d!